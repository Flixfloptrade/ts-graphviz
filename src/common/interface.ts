export interface IDot {
  toDot(): string;
}

export interface IEdgeTarget {
  toEdgeTargetDot(): string;
}
