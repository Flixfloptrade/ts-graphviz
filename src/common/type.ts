// tslint:disable: no-namespace

/**
 * Root cluster type.
 *
 * @description
 * digraph" if you want to use a directional graph.
 *
 * "graph" if you want to use an omnidirectional graph.
 */
export type RootClusterType = 'digraph' | 'graph';
export namespace RootClusterType {
  /**
   * A directional graph type.
   */
  export const digraph: RootClusterType = 'digraph';
  /**
   * An omnidirectional graph type.
   */
  export const graph: RootClusterType = 'graph';
}

/**
 * All of cluster type.
 * @description
 * If you want a hierarchy, use "subgraph".
 */
export type ClusterType = RootClusterType | 'subgraph';
export namespace ClusterType {
  /**
   * A directional graph type.
   */
  export const digraph: ClusterType = 'digraph';
  /**
   * An omnidirectional graph type.
   */
  export const graph: ClusterType = 'graph';
  /**
   * Graph that is not Root cluster.
   *
   * Represents the hierarchy of a graph.
   */
  export const subgraph: ClusterType = 'subgraph';
}

/**
 * Directive indicating which direction the Edge should point.
 */
export type Compass = 'n' | 'ne' | 'e' | 'se' | 's' | 'sw' | 'w' | 'nw' | 'c';
export namespace Compass {
  /** Upper part */
  export const n: Compass = 'n';
  /** Upper left */
  export const ne: Compass = 'ne';
  /** Left part */
  export const e: Compass = 'e';
  /** Lower left */
  export const se: Compass = 'se';
  /** Lower part */
  export const s: Compass = 's';
  /** Lower right */
  export const sw: Compass = 'sw';
  /** Right part */
  export const w: Compass = 'w';
  /** Upper right */
  export const nw: Compass = 'nw';
  /** Center */
  export const c: Compass = 'c';
  export const all: ReadonlyArray<string> = [n, ne, e, se, s, sw, w, nw, c];
}

/**
 * Determine whether the character string satisfies the Compass condition.
 */
export function isCompass(str: string): str is Compass {
  return Compass.all.includes(str);
}
