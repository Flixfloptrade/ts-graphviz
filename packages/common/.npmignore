api-extractor.json
etc
src
tmp
tsconfig.json
tsdoc.json
vite.config.ts
