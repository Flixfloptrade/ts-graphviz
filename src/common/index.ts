export * from './interface';
export * from './abstract';
export * from './util';
export * from './type';
