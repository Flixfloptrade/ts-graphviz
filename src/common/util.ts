/**
 * Type indicating that it is a constructor of T.
 * @hidden
 */
export type Type<T> = new (...args: any[]) => T;
