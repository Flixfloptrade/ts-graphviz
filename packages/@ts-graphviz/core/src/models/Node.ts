import { NodeModel, NodeAttributeKey, NodeAttributesObject, Port, ForwardRefNode, define } from '@ts-graphviz/common';
import { DotObject } from './DotObject.js';
import { AttributesGroup } from './AttributesGroup.js';

/**
 * DOT object class representing a node.
 * @group Models
 */
@define({ type: 'Node' })
export class Node extends DotObject implements NodeModel {
  public comment?: string;

  public readonly attributes: AttributesGroup<NodeAttributeKey>;

  constructor(public readonly id: string, attributes?: NodeAttributesObject) {
    super();
    this.attributes = new AttributesGroup(attributes);
  }

  public port(port: string | Partial<Port>): ForwardRefNode {
    if (typeof port === 'string') {
      return { id: this.id, port };
    }
    return { id: this.id, ...port };
  }
}
