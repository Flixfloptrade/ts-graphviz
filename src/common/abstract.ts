import { IDot } from '../common';
// tslint:disable: max-classes-per-file

/**
 * @category Base
 */
export abstract class GraphvizObject {}

/**
 * @category Base
 */
export abstract class DotBase extends GraphvizObject implements IDot {
  public abstract toDot(): string;
}
