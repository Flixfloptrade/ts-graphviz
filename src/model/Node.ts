import { DotBase } from '../common';
import { IEdgeTarget } from '../common/interface';
import { Attributes } from './attributes';
import { Literal } from './Literal';

/**
 * @category Primary
 */
export class Node extends DotBase implements IEdgeTarget {
  public readonly idLiteral: Literal;
  public readonly attributes = new Attributes();
  constructor(public readonly id: string) {
    super();
    this.idLiteral = new Literal(id);
  }

  public toDot(): string {
    const target = this.idLiteral.toDot();
    const attrs = this.attributes.size > 0 ? ` ${this.attributes.toDot()}` : '';
    return `${target}${attrs};`;
  }

  public toEdgeTargetDot() {
    return this.idLiteral.toDot();
  }

  public port(port: string): NodeWithPort {
    return new NodeWithPort(this, port);
  }
}

// tslint:disable-next-line: max-classes-per-file
export class NodeWithPort implements IEdgeTarget {
  public readonly port: Literal;
  constructor(public readonly node: Node, port: string) {
    this.port = new Literal(port);
  }

  public toEdgeTargetDot() {
    return `${this.node.idLiteral.toDot()}:${this.port.toDot()}`;
  }
}

export type EdgeTargetLike = IEdgeTarget | string;

/**
 * @hidden
 */
export function isEdgeTarget(node: any): node is IEdgeTarget {
  return node instanceof Node || node instanceof NodeWithPort;
}

/**
 * @hidden
 */
export function isEdgeTargetLike(node: any): node is EdgeTargetLike {
  return typeof node === 'string' || isEdgeTarget(node);
}
