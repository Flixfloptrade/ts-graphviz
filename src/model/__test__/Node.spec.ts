import { DotBase, GraphvizObject } from '../../common';
import { Node } from '../Node';

describe('class Node', () => {
  let node: Node;
  beforeEach(() => {
    node = new Node('test');
  });

  it('should be instance of Node/DotBase/GraphvizObject', () => {
    expect(node).toBeInstanceOf(Node);
    expect(node).toBeInstanceOf(DotBase);
    expect(node).toBeInstanceOf(GraphvizObject);
  });

  describe('renders correctly by toDot method', () => {
    it('simple node', () => {
      expect(node.toDot()).toMatchSnapshot();
    });

    it('has some attributes', () => {
      node.attributes.set('label', 'this is test');
      node.attributes.set('color', 'red');
      expect(node.toDot()).toMatchSnapshot();
    });

    describe('label attribute behavior', () => {
      it('plain text label to be quoted by double quotation', () => {
        node.attributes.set('label', 'this is test');
        expect(node.toDot()).toMatchSnapshot();
      });

      it('html like', () => {
        node.attributes.set('label', '<<B>this is test</B>>');
        expect(node.toDot()).toMatchSnapshot();
      });
    });
  });
});
