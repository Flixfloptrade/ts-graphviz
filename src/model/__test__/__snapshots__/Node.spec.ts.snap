// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class Node renders correctly by toDot method has some attributes 1`] = `
"\\"test\\" [
  label = \\"this is test\\",
  color = \\"red\\",
];"
`;

exports[`class Node renders correctly by toDot method label attribute behavior html like 1`] = `
"\\"test\\" [
  label = <<B>this is test</B>>,
];"
`;

exports[`class Node renders correctly by toDot method label attribute behavior plain text label to be quoted by double quotation 1`] = `
"\\"test\\" [
  label = \\"this is test\\",
];"
`;

exports[`class Node renders correctly by toDot method simple node 1`] = `"\\"test\\";"`;
